### PR TITLE
Add support to render components with sidecar template files

### PR DESCRIPTION
**Pull request by joelhawksley**

### CODE REVIEW

File: app/components/issues/badge.html.erb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<%= render(Primer::State.new(color: STATES[state][:color], title: "Status: #{state.to_s.titleize}")) do %>
```
Comment (rmacklin) [inline app/components/issues/badge.html.erb]: When the template was defined in the `issues/badge.rb` file it was easier to tell that `STATES` was referencing the lexically-scoped `Issues::Badge::STATES`.

Moving it to a sidecar file makes that less evident, which makes me wonder if you'd recommend not relying on the lexical scope of the ruby class when using a separate template file (e.g. would it be recommended to replace `STATES` with the fully-qualified `Issues::Badge::STATES` here)?

Comment (joelhawksley) [reply app/components/issues/badge.html.erb]: @rmacklin that's a good point, and I think a fully scoped lookup would be more clear. 

This implementation still allows for templates to be defined inside the component file, as `Box` and `StringInterp` in this repo do. 

The main reason we've switched to sidecars as our default is that a lot of editors didn't play nicely with the HEREDOCs.

I'm guessing which one is used will be a matter of taste.

File: app/lib/action_view/component.rb [hunk 2]
```diff
@@ -4,28 +4,68 @@ class Component < ActionView::Base
...
 
-  private
+    def self.template
+      filename = self.instance_method(:initialize).source_location[0]
+
+      raise NotImplementedError.new("Subclasses of ActionView::Component must implement #initialize") if filename == __FILE__
+
+      filename_without_extension = filename[0..-(File.extname(filename).length + 1)]
+
+      erb_template_path = filename_without_extension+".html.erb"
```
Comment (tomasc) [inline app/lib/action_view/component.rb]: @joelhawksley I see the support for `.html.erb` is hardwired in. Is there a way we could rely on standard Rails template rendering that supports other template languages (such as slim)?

A nice reference can be found in [Dissasembing Rails – Template Rendering](https://medium.com/rubyinside/disassembling-rails-template-rendering-1-51795f579577).

Would that be a next step? Would you be interested in PR that tries to use standard Rails-y template rendering?

Comment (joelhawksley) [reply app/lib/action_view/component.rb]: @tomasc you're 100% right here, this is totally a hack 😄 

Sure, send a PR, as long as you are OK being marked as a co-author if your work is incorporated into any upstream commits to Rails 😄

**Comment by cgriego** [comment]
One of the stated hopes for components was to address the testability and code coverage measurability of these reusable view objects. That hadn't been addressed really, but this feels like a step away from it. Using helpers like `content_tag` inside the Ruby class could work, but what about exposing Builder or something like it to construct HTML inside the class? And there could also be a `render` method available inside the class to render template files to a developer's heart's content.

**Comment by joelhawksley** [comment]
@cgriego this PR doesn't affect the testability, it simply changes the source of the template string.

> That hadn't been addressed really

Can you be more specific? We're able to unit test components now, and are able to at least audit what methods on the component have been called.

I'd love to see a sample implementation using `content_tag` and the associated benchmarks!

File: app/lib/action_view/component.rb (hunk 2 above)
Comment (tomasc) [reply app/lib/action_view/component.rb]: @joelhawksley I started something here https://github.com/joelhawksley/actionview-component-demo/pull/12. However, I could use some help with properly evaluating nested components in the context of the root component. See the PR for more details.

**Comment by tomasc** [comment]
@cgriego I am using components with the `content_tag` for quite some time. There is nothing difficult about that – those can be simple PORO classes.

I think the appeal of @joelhawksley approach is the possibility to use simplified (shorter, more HTML-like) markup – be it `.html.erb` or `.slim` or whatever – and take benefit of standard Rails-y things such as fragment caching.

The missing piece in my opinion are CSS/JS assets located in the same folder as the component, and perhaps even some mechanism for namespacing.

**Pull request merged**